Sources/BSON/Document.swift

```diff
@@ -1,77 +1,156 @@
+
+        for (key, value) in keyValuePairs {
+            self._buffer.writeInteger(value.bsonValue.bsonType.rawValue, as: UInt8.self)
+            self._buffer.writeCString(key)
+            value.bsonValue.write(to: &self._buffer)
+        }
+        // BSON null terminator
+        self._buffer.writeInteger(0, as: UInt8.self)
+
+        guard let byteLength = Int32(exactly: self._buffer.writerIndex - start) else {
+            fatalError("Data is \(self._buffer.writerIndex - start) bytes, "
+                + "but maximum allowed BSON document size is \(Int32.max) bytes")
+        }
+        // Set byteLength in reserved space
+        self._buffer.setInteger(byteLength, at: 0, endianness: .little, as: Int32.self)
+    }
 
-    /// Initializes a new, empty `Document`.
-    public init() { fatalError("Unimplemented") }
+    /// Initializes a new, empty `BSONDocument`.
+    public init() {
+        self.keySet = Set()
+        self._buffer = BSON_ALLOCATOR.buffer(capacity: 5)
+        self._buffer.writeInteger(5, endianness: .little, as: Int32.self)
+        self._buffer.writeBytes([0])
+    }
 
     /**
-     * Initializes a new `Document` from the provided BSON data.
+     * Initializes a new `BSONDocument` from the provided BSON data.
      *
      * - Throws:
      *   - `InvalidArgumentError` if the data passed is invalid BSON
      *
      * - SeeAlso: http://bsonspec.org/
      */
-    public init(fromBSON bson: Data) throws { fatalError("Unimplemented") }
+    public init(fromBSON bson: Data) throws {
+        var buffer = BSON_ALLOCATOR.buffer(capacity: bson.count)
+        buffer.writeBytes(bson)
+        try BSONDocument.validate(buffer)
+        self = BSONDocument(fromUnsafeBSON: buffer)
+    }
+
+    internal init(fromUnsafeBSON bson: Data) {
+        // trust the incoming format
+        var buffer = BSON_ALLOCATOR.buffer(capacity: bson.count)
+        buffer.writeBytes(bson)
+        self = BSONDocument(fromUnsafeBSON: buffer)
+    }
 
     /**
-     * Initializes a new `Document` from the provided BSON data.
+     * Initializes a new `BSONDocument` from the provided BSON data.
      *
      * - Throws:
      *   - `InvalidArgumentError` if the data passed is invalid BSON
      *
      * - SeeAlso: http://bsonspec.org/
      */
-    public init(fromBSON bson: ByteBuffer) throws { fatalError("Unimplemented") }
+    public init(fromBSON bson: ByteBuffer) throws {
+        try BSONDocument.validate(bson)
+        self = BSONDocument(fromUnsafeBSON: bson)
+    }
+
+    internal init(fromUnsafeBSON bson: ByteBuffer) {
+        self.keySet = Set()
+        self._buffer = bson
+        for (key, _) in self {
+            self.keySet.insert(key)
+        }
+    }
 
-    /// The keys in this `Document`.
-    public var keys: [String] { fatalError("Unimplemented") }
+    /// The keys in this `BSONDocument`.
+    public var keys: [String] { self.map { key, _ in key } }
```

Comment:
this `map` decodes the entire document just to get the keys, which we already store in a separate data structure. That structure is undordered though, so I don't think we can use it here. We could possibly replace the `Set` of keys with an `NSMutableOrderedSet` as a performance optimization in the future. That said, perf is out of scope for this work, so I think it's fine as-is.